src/typechecker.erl

```diff
@@ -4812,6 +4819,67 @@ create_fenv(Specs, Funs) ->
       ]
      ).
 
+%% @doc Check for undefined remote types in a separate pass over `Forms'.
+%%
+%% The main typechecker pass defined in `type_check_forms/2' removes position information from types
+%% as soon as they're encountered. It's done to make types occuring on different source
+%% lines represented the same, therefore easy to compare.
+%%
+%% However, in case of references to undefined types (that is, programmer mistakes) it means that
+%% an error which ought to carry this position information to point the user at the source line
+%% to fix is missing.
+%%
+%% This pass over the AST does not remove position information, so invalid remote type references
+%% can be reported properly.
+%%
+%% See https://github.com/josefs/Gradualizer/issues/379 for more information.
+%% @end
+-spec check_undefined_types(Forms) -> Errors when
+      Forms :: gradualizer_file_utils:abstract_forms(),
+      Errors :: [ Error ],
+      Error :: {not_exported, remote_type, erl_anno:anno(), {module(), atom(), non_neg_integer()}}
+             | {undef, remote_type, erl_anno:anno(), {module(), atom(), non_neg_integer()}}.
+check_undefined_types(Forms) ->
+    gradualizer_lib:fold_ast(fun check_remote_type/2, [], Forms).
```

Comment:
Correct me if I'm wrong, but I can't see that we handle the following:

* remote types which expand to structures containing user types or records which are undefined
* remote types which expand to structures containing more remote types
* user types and records belong to a remote module (i.e. they come from expanding a remote type) which refer to other remote types, user types and records
* cycles of the above

Some of it could be solved if gradualizer_db would perform linting when parsing a module, but I don't think it does (yet). (Btw, that's another topic: Unify the gradualizer_db and typechecker code to parse and extracts stuff from a parse tree.)

Reply:
What would the difference between:

> remote types which expand to structures containing user types or records which are undefined

and

> user types and records belong to a remote module (i.e. they come from expanding a remote type) which refer to other remote types, user types and records

be? Could you give an example?

I've added a number of tests - could you confirm if they capture what you had in mind above? Apparently, there was no need to change the check itself.

Reply:
> remote types which expand to structures containing user types or records which are undefined

```Erlang
-module(foo).
-export_type([t/0]).
-type t() :: {undef(), #undef{}}.
```

> user types and records belong to a remote module (i.e. they come from expanding a remote type) which refer to other remote types, user types and records

```Erlang
-module(bar).
-export_type([t/0]).
-type t() :: [w() | r()].
-type w() :: {foo:undef()}.
-record(r, {f :: foo:undef()}).
```

If these are covered by tests, then I suspect the line and column are either zero (because I can't see that they're covered by the new pass, but only by the regular typechecking pass) or if they're non-zero, maybe line and column refer to the location in foo or bar where the error is rather than the module being type checked. We can probably report the line and column where foo:t() or bar:t() is used in the module being checked rather than where the error is in the foo and bar modules.

Reply:
You were right, this work so far only improved one particular situation when the reference was to a remote type which was not exported. All the other reports were still based on type representation fetched from `gradualizer_db` where position information was already lost, so the reported lines were 0.